core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -0,0 +1,755 @@
+ * If the number of enqueued pending write operations becomes equal or larger than the modification queue, then any
+ * subsequent write will be added to the queue, but the returned Stage will not complete until the current batch completes
+ * in an attempt to provide some backpressure to slow writes.
+ * <p>
+ * Read operations may be resolved by this store immediately if the given key is still being updated in the
+ * delegate store or if it is enqueued for the next batch. If the key is in neither it will query the underlying store
+ * to acquire it.
+ * @author wburns
+ * @since 11.0
+ * @param <K> key type for the store
+ * @param <V> value type for the store
+ */
+public class AsyncNonBlockingStore<K, V> extends DelegatingNonBlockingStore<K, V> implements Consumer<Flowable<AsyncNonBlockingStore.Modification>> {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private static final boolean trace = log.isTraceEnabled();
+   private final NonBlockingStore<K, V> actual;
+
+   // Any non-null value can be passed to `onNext` when a new batch should be submitted - A value should only be
+   // submitted to this if the `batchFuture` was null and the caller was able to assign it to a new value
+   private final MulticastProcessor<Object> requestFlowable;
+   // Submit new Modifications to this on every write
+   private volatile FlowableProcessor<Modification> submissionFlowable;
+   // Closing this will stop the subsmissionFlowable subscription
+   private volatile Disposable startSub;
+
+   private Executor nonBlockingExecutor;
+   private int segmentCount;
+   private int modificationQueueSize;
+   private PersistenceConfiguration persistenceConfiguration;
+   private AsyncStoreConfiguration asyncConfiguration;
+
+   // "Non blocking" scheduler used for the purpose of delaying retry batch operations on failures
+   private ScheduledExecutorService scheduler;
+
+   // This variable will be non null if there is a pending batch being sent to the underlying store
+   // If a request causes the modification queue to overflow it will receive a stage back that is only complete
+   // when this future is completed (aka. previous replication has completed)
+   @GuardedBy("this")
+   private CompletableFuture<Void> batchFuture;
+
+   // This variable will be non null if the underlying store has been found to be not available
+   // Note that the async store will still be available as long as the queue size (ie. modificationMap.size) is not
+   // greater than the configured modificationQueueSize
+   @GuardedBy("this")
+   private CompletableFuture<Void> delegateAvailableFuture;
+
+   // Any pending modifications will be enqueued in this map
+   @GuardedBy("this")
+   private Map<Object, Modification> pendingModifications = new HashMap<>();
+   // If there is a pending clear this will be true
+   @GuardedBy("this")
+   private boolean hasPendingClear;
```

Comment:
Intellij is reporting a lot of access outside a `synchronized` block for most of the fields!

Reply:
I just double checked and this field is fine. Maybe the `Modification#apply` is tricking it?

Reply:
I did find a couple in the next two, should be fixed now.

Reply:
yep, `Modification#apply` is still complaining. 
you can add a method in AsyncNonBlockingStore with the `apply()` logic (the logic is similar between put/remove)

Reply:
I am not a fan of changing the code to a do a switch inside another method, where using polymorphism is quite cleaner. I just wish there was a @GuardedBy value where you can specify an argument.

Reply:
oh sorry I misunderstood what you said. Let me do that.

Reply:
Updated, should be satisfied now :)